### PR TITLE
[HSDP] Fix Node 1 unable receive parameters from Node 0

### PR DESCRIPTION
**Pull request by lxg2015**
When use hybrid_shard mode FSDP, 
state.process_group means gpu_0,1,,,~,7 on node 0，so gpus on node 1 cannot receive parameters, setting process_group to default_group（global_group）can fix this issue

Fixes #ISSUE_NUMBER

### CODE REVIEW

**Comment by awgu** [comment]
Thanks for pointing this issue out!

I think that the current broadcast semantics over only the sharding process group is an issue. However, I am not sure if using the global process group is the right thing to do in general.

I think the right semantics are to "union" the sharding process group and replication process group and broadcast over this unioned process group. When only using HSDP alone, this union is the global process group. However, if HSDP does not have full ownership over the cluster (e.g. if composing with some other parallelism), then the union may not be the global process group.

cc: @wanchaol @wz337 Would `DeviceMesh` support something like: We have a 2D submesh from some global mesh (possibly the global mesh _is_ just the 2D mesh), and we call a collective over two dimensions of the mesh? Could `DeviceMesh` initialize new process groups if needed under the hood?

**Comment by lxg2015** [comment]
@awgu  You are right, there is indeed an issue with the global process group. I have modified the logic here. 

Now, I will first synchronize the parameters on gpu0 to gpu1,2,...,7 on  node0, and then according to state._inter_node_pg synchronize the parameters from gpu0 to gpu8, gpu1 to gpu9, gpu2 to gpu10,  .... , so params on gpu0 broadcast to all ranks.

Is this all right ? thxs for your reply

**Comment by lxg2015** [comment]
@awgu I add a unit test at [PR](https://github.com/pytorch/pytorch/pull/108392), and test successfully on 4 GPUs and 8 GPUs.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 1]
```diff
@@ -205,6 +205,48 @@ def test_hsdp_save_load_state_dict(self):
...
+            FSDP,
+            sharding_strategy=ShardingStrategy.HYBRID_SHARD,
+            use_orig_params=True,
+            sync_module_states=True,
+            process_group=(my_shard_group, my_replicate_group),
+        )
+        model = fsdp_ctor(model)
+
+        with FSDP.state_dict_type(model, StateDictType.FULL_STATE_DICT):
+            assert (model.lin1.weight == 0).all()
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:246]: nit: I think the error messaging _might_ be better if we use `self.assertTrue(model.lin1.weight == 0).all()`? Or at least, I think using `self.assert<...>` is the common practice.

File: torch/distributed/fsdp/_init_utils.py [hunk 2]
```diff
@@ -516,6 +516,10 @@ def _init_param_handle_from_module(
         _sync_module_params_and_buffers(
             fully_sharded_module, managed_params, state.process_group
         )
+        if hasattr(state, '_inter_node_pg'):
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:519]: nit: I wonder if this might be brittle for checking if we are using HSDP. Perhaps, we can do `getattr(state, "_inter_node_pg", None) is not None`?

**Review by awgu — APPROVED** [review]
Thank you @lxg2015!

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 1 above)
Comment (lxg2015) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py]: yes, I have changed assert to self.assertTrue

File: torch/distributed/fsdp/_init_utils.py (hunk 2 above)
Comment (lxg2015) [reply torch/distributed/fsdp/_init_utils.py]: yes, I have changed this

**Comment by lxg2015** [comment]
Hello @awgu , can this PR be merged? In order to reduce memory consumption, many other packages will only load the checkpoint on gpu 0, like accelerate. if we don't fix it, this will cause abnormal loss.

**Comment by lxg2015** [comment]
Hi @awgu，I just replace getattr(state, "_inter_node_pg", None) with state.sharding_strategy to revent getattr being repeatedly called, please merge again

**Comment by wz337** [comment]
> Thanks for pointing this issue out!
> 
> I think that the current broadcast semantics over only the sharding process group is an issue. However, I am not sure if using the global process group is the right thing to do in general.
> 
> I think the right semantics are to "union" the sharding process group and replication process group and broadcast over this unioned process group. When only using HSDP alone, this union is the global process group. However, if HSDP does not have full ownership over the cluster (e.g. if composing with some other parallelism), then the union may not be the global process group.
> 
> cc: @wanchaol @wz337 Would `DeviceMesh` support something like: We have a 2D submesh from some global mesh (possibly the global mesh _is_ just the 2D mesh), and we call a collective over two dimensions of the mesh? Could `DeviceMesh` initialize new process groups if needed under the hood?

Yes. I think you can call a collective over two dimensions of the mesh. I believe Wanchao removes DeviceMesh's collective, since they were just a thin layer of functional collective. You should be able to use functional collective directly for this. If I understand your use case correctly, this may be something you are looking for. Code pointer: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py#L161 

And Yes, DeviceMesh would initialize new process group if needed. The current logic is that for mesh same as the world size, it will re-use if it has been initialized. For sub pgs, it will go throught the creation. See pointer: 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/device_mesh.py#L213 

cc. @wanchaol

**Comment by lxg2015** [comment]
Hi @awgu, All checks appear to be fine, can this PR be merged？:grin: Or is DeviceMesh a better way to fix

**Comment by haichaoyu** [comment]
Hi @lxg2015, one question about this PR fix.

[Here](https://github.com/lxg2015/pytorch/blob/ac714a3abebaccf41d53ec77609f8a5bedd60f77/torch/distributed/fsdp/_init_utils.py#L570), `_get_orig_params` is called, and `_get_orig_params` does not return `FlatParameter`'s. When some [auto_wrap_policy](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel) is applied, there will be some `FlatParameter`'s due to nested wrapping. In this case, not all model parameters will be successfully synced to make them consistent? Thanks!

**Comment by awgu** [comment]
@haichaoyu Each `FullyShardedDataParallel` instance should be responsible for syncing its own managed parameters, so you would want to pass `sync_module_states=True` for all `FullyShardedDataParallel` instances to broadcast the entire model from rank 0.

For example, suppose you had a transformer model with transformer blocks, where you apply FSDP to each transformer block and then finally to the root transformer. When you apply FSDP to each transformer block with `sync_module_states=True`, it broadcasts the transformer block's parameters from rank 0. Finally, when you apply FSDP to the root, it broadcasts the root's parameters (e.g. embedding weight, output projection weight) from rank 0 and does not re-broadcast the already flattened transformer block parameters.

Let me know if this makes sense!

**Comment by haichaoyu** [comment]
Got it. Thanks for detailed explanation!

**Comment by haichaoyu** [comment]
@awgu Another relevant question.
In the second time of calling [_sync_module_params_and_buffers](https://github.com/pytorch/pytorch/pull/108331/files#diff-ba816a28d6735a271848b181ae64fc6278d6e818b27f6037f9ca926cbfd9fb67R576), [`buffer.FSDP_SYNCED`](https://github.com/lxg2015/pytorch/blob/ac714a3abebaccf41d53ec77609f8a5bedd60f77/torch/distributed/fsdp/_init_utils.py#L1043) is already set `True` and will not be synced between nodes. Does this cause inter-node inconsistency issue if model has randomly initialized buffers? Thanks!

**Comment by awgu** [comment]
If a parent module re-initializes the buffer of a child module, where the child module is part of a different FSDP wrapping, then yes, this could cause issues. The general guidance for FSDP is that each module should only initialize its directly owned parameters/buffers to avoid cases like this.
